Get proper mFeature in relation reference widget

signedav opened the pull request:
The issue was, that the embedded attribute form of the relation reference didn't got updated in case there was the first entry in the combo box selected (because feature was not valid).
With this fix we update the embedded form after the initialization.

IMHO this should have been needed before and not only for the first entry in the combo box. But because the combo box gets an update in case it's not the first entry, the signal is fired to update the combo box and so the embedded form. In the older version it seems the combo box got rearranged even when it's the first entry, and so it worked back then.

Anyway the embedded form is updated (and the related feature highlighted) now in any case after the initialization.

Fix #19342

signedav committed:
Get mFeature before updating attributeform

The issue was, that the attributeform didn't got updated in case there was the first entry in the combobox selected.
If there was not the first combobox entry selected, there has been a setindex after the initialization and the attributeform got updated.
But the first entry needs this call.

Fix #19342

signedav committed:
highlight on initialization

m-kuhn commented on src/gui/editorwidgets/qgsrelationreferencewidget.cpp:
```diff
@@ -555,6 +555,9 @@ void QgsRelationReferenceWidget::init()
     // Only connect after iterating, to have only one iterator on the referenced table at once
     connect( mComboBox, static_cast<void ( QComboBox::* )( int )>( &QComboBox::currentIndexChanged ), this, &QgsRelationReferenceWidget::comboReferenceChanged );
+
+    mReferencedLayer->getFeatures( mComboBox->currentFeatureRequest() ).nextFeature( mFeature );
```
Having unconditional I/O in `init()` looks like a potential performance killer, a solution that works with the feedback from the worker thread would be preferrable.

signedav replied:
I understand this input and I'm looking for a better solution. 
But when we remove this part, shouldn't the following line `updateAttributeEditorFrame( mFeature );` be removed as well? Because the `mFeature` is not valid? Or how was that supposed to work?

signedav committed:
emit currentIndexChanged onDataChanged

Because this effects that the data related to the combobox (emedded forms) got updated. they update because of an index change, but needed to be updated even when the index does not change (but the data). This goes over the same canal here.

signedav commented:
@m-kuhn last commit would resolve the problem as well. Started over the `dataChanged` signal when loading combo box with `QgsFeatureFilterModel::updateCompleter()`...

signedav commented:
Anyway I don't really like that solution because it simulates an `indexChange()` though the index has always been 0. What really has to be done is, that the embedded form gets loaded on loading the form/tab - and this would be on the `init()` - so what do you think about the originally by you proposed option to set a condition on updating the embedded form? Like:
```
    if( mEmbedForm )
    {
      mReferencedLayer->getFeatures( mComboBox->currentFeatureRequest() ).nextFeature( mFeature );
      highlightFeature( mFeature );
      updateAttributeEditorFrame( mFeature );
    }
```
Only weakness there is, that the related item is not highlighted when no embedded form is active on selecting the tab (only highlighted on opening the feature form).

m-kuhn commented:
I think where the real problem is, is that at init time the index never changes from "invalid" to 0. Maybe there needs to be a *bool mInitialized*, that indicates if we have at least once emitted a signal.

signedav committed:
initialization of embeddet form only once

signedav commented:
Hi @pblottiere 
Would you mind to review this, because Matthias is not available?

I solved it with the original approach but with the flag `mFeatureInitialized` - so it's not initialized every time when the form is loaded.

signedav committed:
fix spelling

m-kuhn commented on src/gui/editorwidgets/qgsrelationreferencewidget.cpp:
```diff
@@ -555,7 +555,13 @@ void QgsRelationReferenceWidget::init()
+    if ( !mFeatureInitialized )
+    {
+      //call it for the first initialization
+      emit mComboBox->currentIndexChanged( mComboBox->currentIndex() );
```
Is this triggering the `getFeature()` in the slot?
In this case it's executed on the main thread anyway, so doesn't make a big difference to the previous approach.

I would propose the following approach:

 * Use this approach with emitting the signal (this keeps code compact in the slot)
 * Remove the initialized condition here
 * Guard the whole `init()` function with a `mInitialized` variable (and not only the emission of this signal)

```
void init()
{
   if ( mInitialized )
     return;

  ...
  mInitialized = true;
}
```

signedav committed:
use initialized for complete init() function

not only for the feature part

m-kuhn reviewed (approved):
Looks good to me now (except for the sip file, but I guess travis will take care of reporting that one)
